### PR TITLE
Allow shortcut representation for buckets

### PR DESCRIPTION
**Pull request by josevalim**

### CODE REVIEW

File: lib/telemetry_metrics.ex [hunk 1]
```diff
@@ -319,12 +319,16 @@ defmodule Telemetry.Metrics do
   Distribution metric builds a histogram of selected measurement's values. Because of that, it is
   required that you specify the histograms buckets via `:buckets` option.
 
-  For example, given `buckets: [0, 100, 200]`, the distribution metric produces four values:
+  TThe buckets is either a list of integers, such as `[100, 200, 300]`, or a two-element tuple,
```
Comment (hauleth) [inline lib/telemetry_metrics.ex]: Typo `TThe`.

File: lib/telemetry_metrics.ex [hunk 2]
```diff
@@ -319,12 +319,16 @@ defmodule Telemetry.Metrics do
...
+  containing the range as first element and the step as second, such as `{100..300, 100}`, which
+  emits the same buckets as `[100, 200, 300]`.
+
+  Given `buckets: [100, 200, 300]`, the distribution metric produces four values:
 
     * number of measurements less than or equal to 0
```
Comment (hauleth) [inline lib/telemetry_metrics.ex]: Is this still true? It would be strange to `[100]` also count measurements less than `0`.

**Comment by josevalim** [comment]
@hauleth fixed both, thanks! :heart:

**Comment by bryannaegele** [comment]
This is very much needed!

When it comes to bucketing, most of the advice tends to avoid fixed size buckets, but they often suggest specific strategies. I'd like to see us add support for adding multiple strategies people could employ versus just fixed size. In order to do so, we'd need a way to explicitly note the strategy by name.

My suggestion for this example would be to make it easy to provide multiple strategies over time:

```elixir
{:fixed_width, [range: 100..200, size: 100]}
```

What do y'all think?

**Comment by josevalim** [comment]
@bryannaegele OpenCensus also only allow fixed buckets, so I would wait a bit more before diverging the  projects further. I also think the backends that work with dynamic buckets require them to be specified in their own UI, rather than upfront?

**Comment by josevalim** [comment]
:heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:

**Comment by bryannaegele** [comment]
@josevalim I'm not sure where that information came from. The bucket sizes have to be declared up front since the measurements are bucketed and discarded as part of aggregation when collecting, so you cannot re-bucket them at a later time. There are examples of non-fixed size buckets in OpenCensus here https://opencensus.io/stats/view/#aggregations and https://opencensus.io/quickstart/erlang/metrics/.

There is a type of histogram that does what you're describing - HDRHistogram, but that requires keeping all of the measurements.

The buckets shown in the Erlang sample are representative of the default Prometheus bucket sizes I believe. Should we continue this conversation in an issue?

**Comment by josevalim** [comment]
@bryannaegele we can continue it here if you don't mind.

>  There are examples of non-fixed size buckets in OpenCensus here https://opencensus.io/stats/view/#aggregations

I am not that well-versed in OpenCensus, which is probably from where the misunderstanding comes from, but I cannot see non-fixed size buckets in the link above. All of the examples in there define the buckets upfront for distributions. I think I am missing something obvious. Help? :)

**Comment by josevalim** [comment]
PS: Opening a new issue is totally fine too.

**Comment by bryannaegele** [comment]
Need to have my coffee ;)

The buckets are of a fixed size, but the _width_ of the buckets aren't uniform.

Example:
```erlang
latency_distribution() ->
    {oc_stat_aggregation_distribution, [{buckets, [0, 25, 50, 75, 100, 200, 400,
                                                   600, 800, 1000, 2000, 4000, 6000]}]}.
```

So given a range and a single divisor, this example cannot be derived. Prometheus has some defaults which also don't neatly fit this pattern https://github.com/prometheus/client_golang/blob/master/prometheus/histogram.go#L61.

If you examine the Go implementation further, there are a number of different strategies which can be used, including the fixed width strategy this PR implements, but also linear, exponential, etc.

https://github.com/prometheus/client_golang/blob/master/prometheus/histogram.go#L74

It would be very helpful I think to support these strategies and also allow the ability to pass a function for custom strategies.

**Comment by josevalim** [comment]
Oh, we support those today. You can always pass a list with the buckets
explicitly. This API is only meant to be a shortcut for the common case.

We could have shorter APIs for the other cases but I wonder how beneficial
those would be compared to explicitly listing them. I probably would
explicitly list the buckets for anything non trivial.
-- 


*José Valimwww.plataformatec.com.br
<http://www.plataformatec.com.br/>Founder and Director of R&D*

**Comment by bryannaegele** [comment]
We can externally declare these strategies to produce these, which is fine for some very custom ones (we have some at Weedmaps) but I'm wondering if we want to force users to implement for these basic ones.

Another option is to just leave that possibility to Reporters to expose those types of things but that then ties these definitions to a single Reporter, thus limiting the portability factor Telemetry.Metrics has.

**Comment by josevalim** [comment]
Exactly. In a nutshell, this PR is not adding anything new beyond a
shortcut and we could add other shortcuts in the future but I am still
unsure they would be beneficial over explicitly listing the buckets out.

For an actual different behavior, then we would need a long conversation to
justify moving away from what OpenCensus uses.

Btw, how would we achieve our current summary in OpenCensus?
-- 


*José Valimwww.plataformatec.com.br
<http://www.plataformatec.com.br/>Founder and Director of R&D*

**Comment by hauleth** [comment]
I think it would be better to just provide “generator functions” that would return list with properly distributed values instead of adding new syntax for specifying buckets. This would provide better “building blocks” and would allow to compose different distributions, for example:

    log(0..100, 10) ++ linear(100..1000, 100) ++ exp(1000..10_000, 5)

Which would create 10 steps in logarithmic scale, then 100 in linear and 5 in exponential. 

--
Łukasz Jan Niemier

**Comment by arkgil** [comment]
> Btw, how would we achieve our current summary in OpenCensus?

I don't think that OpenCensus has anything like that today. The original set of 4 metric types was based directly on OpenCensus. The summary was inspired by Prometheus. The summary could create multiple OpenCensus metrics (e.g. a counter and a histogram), but I'm not sure if that's beneficial.

**Pull request merged**